build/functions.j2

```diff
@@ -184,13 +184,24 @@ function {{ "scan_" ~ function if scan else function }}() {
         bash "${hook}"
         return=$?
         if [[ "${return:-1}" != 0 ]]; then
-          popd > /dev/null
-          if [[ "${FAIL_FAST:-false}" == "true" ]]; then
-            _feedback INFO "FAIL_FAST is set to ${FAIL_FAST}; returning the exit code of ${return} immediately"
-            return "${return}"
+          message="${hook} failed to execute properly"
+
+          if [[ "${LEARNING_MODE,,}" != "true" ]]; then
+            # easy_infra denied running the hook because the script failed
+            _log "easy_infra.stdouterr" denied failure "easy_infra" "${dir}" string "${message}"
+            _feedback DEBUGGING "${hook} exited ${return}; learning mode was ${LEARNING_MODE:-null or unset}. Returning ${return}"
+            popd > /dev/null
+            if [[ "${FAIL_FAST:-false}" == "true" ]]; then
+              _feedback INFO "FAIL_FAST is set to ${FAIL_FAST}; returning the exit code of ${return} immediately"
+              return "${return}"
+            else
+              dir_exit_codes["${dir}"]="${return}"
+              continue
```

Comment:
Based on this, if `FAIL_FAST` is `false` wouldn't it never actually fail?  It just populates the `dir_exit_codes` and allows the logic to go past the hooks stage.  I expected we would let all the hooks run, but then fail before continuing, if `FAIL_FAST` was false.  Please reflect in tests as well to ensure this understanding is accurate